### PR TITLE
Add UI Tests for Block, Form, Report and List

### PR DESCRIPTION
**Pull request by MedAliMarz**
This PR adds missing UI tests for:
- Block
- Form
- Report
- List

### CODE REVIEW

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 1]
```diff
@@ -80,6 +83,105 @@ class ReportTests: GaliteVUITestBase() {
...
+    _expectOne<DReport>()
+    simpleReport.csv.triggerCommand()
+
+    _expectOne<DownloaderDialog>()
+    val downloadDialog = _get<DownloaderDialog>()
+
+    assertTrue(downloadDialog.isOpened)
+    val anchor = downloadDialog._get<Anchor> {  }
+
+    assertTrue(anchor.href.contains(simpleReport.title))
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt:107]: you can assert also on` div with class='download-file-title' `end with CSV (add it to pdf, xls and xlsx tests)

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 2]
```diff
@@ -80,6 +83,105 @@ class ReportTests: GaliteVUITestBase() {
     }
   }
 
+  @Test
+  fun `test export CSV`() {
+    // Login
+    login()
+
+    // Opens a form that contain a report command
+    formWithReport.open()
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt:92]: remove this part from your tests and add : 
```
@Before
  fun `login to the App`() {
    login()
    formWithReport.open()
  }
```

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 3]
```diff
@@ -80,6 +83,105 @@ class ReportTests: GaliteVUITestBase() {
...
+
+    // Trigger the report command
+    formWithReport.report.triggerCommand()
+
+    // Check that the report is displayed
+    _expectOne<DReport>()
+    simpleReport.xls.triggerCommand()
+    // Trigger the command and check that the downloaderDialog is opened
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt:125]: put comment before triggerCommand()

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsForm.kt [hunk 4]
```diff
@@ -50,8 +52,51 @@ class TestFieldsForm : DictionaryForm() {
...
+    label = "Update Mode",
+    help = "Change the mode to update"
+  ) {
+    icon = "edit"
+  }
+
   val blockWithDifferentTypes = insertBlock(BlockWithDifferentTypes())
-  val blockWithAllFieldVisibilityTypes = insertBlock(BlockWithAllFieldVisibilityTypes())
+  val blockWithAllFieldVisibilityTypes = insertBlock(BlockWithAllFieldVisibilityTypes()) {
+    command(item = query){
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsForm.kt:84]: add space before braces, (also for update and insert)

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/list/ListTests.kt [hunk 5]
```diff
@@ -91,6 +94,78 @@ class ListTests: GaliteVUITestBase() {
     assertEquals("informations training 2", formWithList.block.informations.findField().value)
   }
 
+  @Test
+  fun `test list command with reordered columns`() {
+    // Login
+    login()
+
+    // Opens a form that contain a list command
+    formWithList.open()
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/list/ListTests.kt:103]: add : 
 ```
@Before
  fun `login to the App`() {
    login()
    formWithList.open()
  }
```

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 6]
```diff
@@ -80,6 +83,105 @@ class ReportTests: GaliteVUITestBase() {
...
+    // Opens a form that contain a report command
+    formWithReport.open()
+
+    // Trigger the report command
+    formWithReport.report.triggerCommand()
+
+    // Check that the report is displayed
+    _expectOne<DReport>()
+    simpleReport.xlsx.triggerCommand()
+    // Trigger the command and check that the downloaderDialog is opened
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt:150]: reverse

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 7]
```diff
@@ -80,6 +83,105 @@ class ReportTests: GaliteVUITestBase() {
...
+    // Opens a form that contain a report command
+    formWithReport.open()
+
+    // Trigger the report command
+    formWithReport.report.triggerCommand()
+
+    // Check that the report is displayed
+    _expectOne<DReport>()
+    simpleReport.pdf.triggerCommand()
+    // Trigger the command and check that the downloaderDialog is opened
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt:175]: reverse

**Review by h-haddad — CHANGES_REQUESTED** [review]

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/block/SimpleBlockTests.kt [hunk 8]
```diff
@@ -87,6 +92,37 @@ class SimpleBlockTests: GaliteVUITestBase() {
...
+    block.visitField.click()
+    // trigger a command
+    testFieldsForm.insert.triggerCommand()
+
+    val dBlock = block.findBlock()
+
+    // verify that the mode have changed in the block
+    assertEquals(Mode.INSERT.value,dBlock.model.getMode())
+    // verify the modification in the visit fields access
+    assertEquals(VConstants.ACS_SKIPPED, block.visitFieldToSkippedField.access[Mode.INSERT.value])
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/block/SimpleBlockTests.kt:114]: I think that will be needed if the access array is unaccessible, @hfazai what do you think, do we add the getter? or directly use the access array?

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/block/SimpleBlockTests.kt [hunk 9]
```diff
@@ -87,6 +92,37 @@ class SimpleBlockTests: GaliteVUITestBase() {
...
+    assertEquals(VConstants.ACS_HIDDEN, block.visitFieldToHiddenField.access[Mode.INSERT.value])
+    assertEquals(VConstants.ACS_MUSTFILL, block.visitFieldToMustFillField.access[Mode.INSERT.value])
+    // verify the modification in the mustfill fields access
+    assertEquals(VConstants.ACS_SKIPPED, block.mustFillToSkippedField.access[Mode.INSERT.value])
+    assertEquals(VConstants.ACS_VISIT, block.mustFillToVisitField.access[Mode.INSERT.value])
+    assertEquals(VConstants.ACS_HIDDEN, block.mustFillToHiddenField.access[Mode.INSERT.value])
+    // verify the modification in the skipped fields access
+    assertEquals(VConstants.ACS_HIDDEN, block.skippedToHiddenField.access[Mode.INSERT.value])
+    assertEquals(VConstants.ACS_VISIT, block.skippedToVisitField.access[Mode.INSERT.value])
+    assertEquals(VConstants.ACS_MUSTFILL, block.skippedToMustFillField.access[Mode.INSERT.value])
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/block/SimpleBlockTests.kt:124]: Added it only for the update button, but not for the query button as the fields access didn't change (even after adding onQueryX, where X is the original field access)

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/form/MultipleBlockFormTests.kt [hunk 10]
```diff
@@ -238,6 +306,56 @@ class MultipleBlockFormTests : GaliteVUITestBase() {
     assertEquals(1, block.editor.item.record)
   }
 
+  @Test
+  fun `test the shortcut help button`() {
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/form/MultipleBlockFormTests.kt:310]: Changed it to `test the actors shortcut dialog`.

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/form/MultipleBlockFormTests.kt [hunk 11]
```diff
@@ -238,6 +306,56 @@ class MultipleBlockFormTests : GaliteVUITestBase() {
...
+    val visibleForm = _get<DForm> {  }
+
+    assertEquals(form, visibleForm)
+    assertFails { multipleBlockSaveForm.findForm() }
+
+    // get back to multipleBlockSaveForm
+    windowsDiv._click()
+    _expect<EnhancedDialog> {  }
+    windowsContainer = _get<VerticalLayout> { classes = "window-items-container" }
+    windowsItems = windowsContainer._find<VWindowsMenuItem> { classes = "item" }
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/form/MultipleBlockFormTests.kt:353]: The windows Container and Items stays the same (same element in the dom) also the `windowsItems[1].click()` is simulating the click even if the `EnhanceDialog` is hidden

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/list/ListTests.kt [hunk 12]
```diff
@@ -91,6 +94,78 @@ class ListTests: GaliteVUITestBase() {
...
+
+    // Check that the grid data is correct
+    grid.expectRows(data.size)
+    data.forEachIndexed { index, it ->
+      grid.expectRow(index, *it)
+    }
+  }
+
+  @Test
+  fun `test list command with sorted rows`() {
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/list/ListTests.kt:134]: Yes, but it should exist, to have a corresponding behavior to Kopi, marked as ignored for now!

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/form/MultipleBlockFormTests.kt [hunk 13]
```diff
@@ -238,6 +306,56 @@ class MultipleBlockFormTests : GaliteVUITestBase() {
...
+    _expect<Dialog> {  }
+    val dialog = _get<Dialog> {}
+
+    assertNotNull(dialog)
+
+    val navigationColumns = _find<VerticalLayout> { classes= "actor-navigationColumn" }
+
+    assertEquals(2, navigationColumns.size)
+    assertEquals(2, navigationColumns[0].children.count())
+    assertEquals(10, navigationColumns[1].children.count())
```
[the remark this answers is not in the document]
Comment (MedAliMarz) [reply galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/form/MultipleBlockFormTests.kt:324]: Done for both, thanks for the suggestion

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    label = "skipped field to mustfill"
+    help = "skipped field"
+    onInsertMustFill()
+    onUpdateMustFill()
+    onQuerySkipped()
+  }
+}
+
+fun main() {
+  runForm(formName = TestFieldsForm())
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt:176]: change to TestFieldsVisibilityForm

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+    label = "hidden field"
+    help = "hidden field"
+  }
+
+  val visitField = visit(domain = INT(3), position = at(1, 1)) {
+    label = "visit field"
+    help = "visit field"
+    minValue = 10
+    maxValue = 50
+  }
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt:90]: unused

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    help = "visit field"
+    onInsertMustFill()
+    onUpdateMustFill()
+    onQueryVisit()
+  }
+
+  val mustFillField = mustFill(domain = STRING(50), position = at(2, 1)) {
+    label = "mustFill field"
+    help = "mustFill field"
+  }
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt:119]: unused

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+    help = "mustFill field"
+    onInsertHidden()
+    onUpdateHidden()
+    onQueryMustFill()
+  }
+
+  val skippedField = skipped(domain = STRING(50), position = at(3, 1)) {
+    label = "skipped field"
+    help = "skipped field"
+  }
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/examples/TestFieldsVisibilityForm.kt:148]: unused

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/list/ListTests.kt [hunk 18]
```diff
@@ -39,23 +39,25 @@ import com.github.mvysny.kaributools.setSortOrder
 import com.vaadin.flow.component.grid.Grid
 import com.vaadin.flow.component.grid.GridSortOrder
 import com.vaadin.flow.data.provider.SortDirection
+import org.junit.Before
+import org.junit.Ignore
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/list/ListTests.kt]: reorder

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 19]
```diff
@@ -39,6 +39,9 @@ import com.github.mvysny.kaributesting.v10.expectRow
 import com.github.mvysny.kaributesting.v10.expectRows
 import com.vaadin.flow.component.grid.Grid
 import com.vaadin.flow.component.html.Anchor
+import com.vaadin.flow.component.html.Div
+import org.junit.Before
+import org.kopi.galite.testing.waitAndRunUIQueue
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt]: reorder

File: galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt [hunk 20]
```diff
@@ -85,102 +88,94 @@ class ReportTests: GaliteVUITestBase() {
 
   @Test
   fun `test export CSV`() {
```
Comment (h-haddad) [inline galite-tests/src/test/kotlin/org/kopi/galite/tests/ui/vaadin/report/ReportTests.kt:90]: check if test export XLSX, test export XLS, test export CSV, test export PDF stay valid after merge with master

**Review by h-haddad — CHANGES_REQUESTED** [review]

**Review by hfazai — APPROVED** [review]

**Pull request merged**